rust/gui-client/src-tauri/src/client/deep_link/linux.rs

```diff
@@ -13,16 +13,29 @@ pub(crate) struct Server {
     listener: UnixListener,
 }
 
+fn sock_path() -> Result<PathBuf> {
+    Ok(known_dirs::runtime()
+        .context("Couldn't find runtime dir")?
+        .join(SOCK_NAME))
+}
+
 impl Server {
     /// Create a new deep link server to make sure we're the only instance
     ///
     /// Still uses `thiserror` so we can catch the deep_link `CantListen` error
     pub(crate) fn new() -> Result<Self, super::Error> {
-        let dir = known_dirs::runtime().context("couldn't find runtime dir")?;
-        let path = dir.join(SOCK_NAME);
-        // TODO: This breaks single instance. Can we enforce it some other way?
+        let path = sock_path()?;
+        let dir = path
+            .parent()
+            .context("Impossible, socket path should always have a parent")?;
+
+        // If the connection succeeds, that means another process has bound the socket,
+        // and we're the 2nd instance, so we should exit
+        if std::os::unix::net::UnixStream::connect(&path).is_ok() {
+            return Err(super::Error::CantListen);
+        }
```

Comment:
I think I'm missing something -- wouldn't we fail this if it fails to bind to the socket, not succeeds?

Reply:
I reworded it - https://github.com/firezone/firezone/pull/4890/files#diff-1c34fff8b944d6df5027d77dd576c1ec9121d0b1ac387da9b9bf0a4fddb241c4R32-R40

We're not binding here, we're connecting. If the connection succeeds, someone else has already bound the socket.

The reason I didn't rely on bind alone is that bind seemed to fail if another process left a socket behind when it exited, but maybe I'm wrong about that 🤔 

There is definitely a TOCTOU logic error in this section, so I also marked that. If I do more testing with Unix domain sockets I can probably simplify and correct all this later. For now I'm okay with it being tech debt.